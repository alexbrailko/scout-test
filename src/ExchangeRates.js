import React, { Component } from 'react';
import { connect } from "react-redux";
import Modal from 'react-modal';
import { resetError } from "./actions/ratesActions";

export class ExchangeRates extends Component {

  state = {
    showModal: true
  };

  handleCloseModal = () => {
    this.setState({ showModal: false });
    this.props.resetError();
  }

  render() {
    const { rates, error, loading, ratesFetched } = this.props;

    if (error) {
      return <Modal
        isOpen={true}
        ariaHideApp={false}
      >
        <div className="modalContent">
          {error.message}
          <button onClick={this.handleCloseModal}>Close Modal</button>
        </div>
      </Modal>
    }

    if (loading) {
      return <div className="loading">Loading...</div>;
    }

    if (ratesFetched) {
      return (
        <table>
          <thead>
            <tr>
              <td><strong>{rates.base} - {rates.date}</strong></td>
            </tr>
          </thead>
          <tbody>
            {Object.entries(rates.rates).map(([key, rate]) =>
              <tr key={key}>
                <td>{key}</td>
                <td>{rate}</td>
              </tr>
            )}
          </tbody>
        </table>
      );
    } else {
      return null;
    }
  }
};

const mapStateToProps = ({rates}) => ({
  rates: rates.items,
  loading: rates.loading,
  error: rates.error,
  ratesFetched: rates.ratesFetched
});

export default connect(mapStateToProps, { resetError })(ExchangeRates);