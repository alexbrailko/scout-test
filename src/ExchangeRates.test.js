import React from 'react';
import { shallow } from 'enzyme';

import { ExchangeRates } from './ExchangeRates';
import Modal from 'react-modal';

const mockData = { base: "EUR", date: "2018-11-07", rates: { AUD: 1.575, BGN: 1.9558 } };

it('should show modal on error', () => {
  const wrapper = shallow(<ExchangeRates error="Error" />);
  expect(wrapper.find(Modal).length).toEqual(1);
  expect(wrapper.find(".modalContent").exists()).toEqual(true);
}); 

it('should close modal on button click', () => {
  const resetError = jest.fn();
  const wrapper = shallow(<ExchangeRates error="Error" resetError={resetError} />);
  wrapper.find('.modalContent button').simulate('click');
  expect(resetError).toHaveBeenCalled();
  expect(wrapper.state('showModal')).toEqual(false);
}); 

it('should show loading when loading data', () => {
  const wrapper = shallow(<ExchangeRates loading={true} />);
  expect(wrapper.find('.loading').length).toEqual(1);
});

it('should render Exchange Rate data', () => {
  const wrapper = shallow(<ExchangeRates ratesFetched={true} rates={mockData} />);
  expect(wrapper).toMatchSnapshot();
}); 

 
