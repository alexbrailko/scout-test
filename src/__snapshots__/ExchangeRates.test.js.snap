// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Exchange Rate data 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ExchangeRates
    rates={
      Object {
        "base": "EUR",
        "date": "2018-11-07",
        "rates": Object {
          "AUD": 1.575,
          "BGN": 1.9558,
        },
      }
    }
    ratesFetched={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <thead>
          <tr>
            <td>
              <strong>
                EUR
                 - 
                2018-11-07
              </strong>
            </td>
          </tr>
        </thead>,
        <tbody>
          <tr>
            <td>
              AUD
            </td>
            <td>
              1.575
            </td>
          </tr>
          <tr>
            <td>
              BGN
            </td>
            <td>
              1.9558
            </td>
          </tr>
        </tbody>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <tr>
            <td>
              <strong>
                EUR
                 - 
                2018-11-07
              </strong>
            </td>
          </tr>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <td>
              <strong>
                EUR
                 - 
                2018-11-07
              </strong>
            </td>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <strong>
                EUR
                 - 
                2018-11-07
              </strong>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "EUR",
                  " - ",
                  "2018-11-07",
                ],
              },
              "ref": null,
              "rendered": Array [
                "EUR",
                " - ",
                "2018-11-07",
              ],
              "type": "strong",
            },
            "type": "td",
          },
          "type": "tr",
        },
        "type": "thead",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <tr>
              <td>
                AUD
              </td>
              <td>
                1.575
              </td>
            </tr>,
            <tr>
              <td>
                BGN
              </td>
              <td>
                1.9558
              </td>
            </tr>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "AUD",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <td>
                  AUD
                </td>,
                <td>
                  1.575
                </td>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "AUD",
                },
                "ref": null,
                "rendered": "AUD",
                "type": "td",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": 1.575,
                },
                "ref": null,
                "rendered": 1.575,
                "type": "td",
              },
            ],
            "type": "tr",
          },
          Object {
            "instance": null,
            "key": "BGN",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <td>
                  BGN
                </td>,
                <td>
                  1.9558
                </td>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "BGN",
                },
                "ref": null,
                "rendered": "BGN",
                "type": "td",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": 1.9558,
                },
                "ref": null,
                "rendered": 1.9558,
                "type": "td",
              },
            ],
            "type": "tr",
          },
        ],
        "type": "tbody",
      },
    ],
    "type": "table",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <thead>
            <tr>
              <td>
                <strong>
                  EUR
                   - 
                  2018-11-07
                </strong>
              </td>
            </tr>
          </thead>,
          <tbody>
            <tr>
              <td>
                AUD
              </td>
              <td>
                1.575
              </td>
            </tr>
            <tr>
              <td>
                BGN
              </td>
              <td>
                1.9558
              </td>
            </tr>
          </tbody>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <tr>
              <td>
                <strong>
                  EUR
                   - 
                  2018-11-07
                </strong>
              </td>
            </tr>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <td>
                <strong>
                  EUR
                   - 
                  2018-11-07
                </strong>
              </td>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <strong>
                  EUR
                   - 
                  2018-11-07
                </strong>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "EUR",
                    " - ",
                    "2018-11-07",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  "EUR",
                  " - ",
                  "2018-11-07",
                ],
                "type": "strong",
              },
              "type": "td",
            },
            "type": "tr",
          },
          "type": "thead",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <tr>
                <td>
                  AUD
                </td>
                <td>
                  1.575
                </td>
              </tr>,
              <tr>
                <td>
                  BGN
                </td>
                <td>
                  1.9558
                </td>
              </tr>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "AUD",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <td>
                    AUD
                  </td>,
                  <td>
                    1.575
                  </td>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "AUD",
                  },
                  "ref": null,
                  "rendered": "AUD",
                  "type": "td",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": 1.575,
                  },
                  "ref": null,
                  "rendered": 1.575,
                  "type": "td",
                },
              ],
              "type": "tr",
            },
            Object {
              "instance": null,
              "key": "BGN",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <td>
                    BGN
                  </td>,
                  <td>
                    1.9558
                  </td>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "BGN",
                  },
                  "ref": null,
                  "rendered": "BGN",
                  "type": "td",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": 1.9558,
                  },
                  "ref": null,
                  "rendered": 1.9558,
                  "type": "td",
                },
              ],
              "type": "tr",
            },
          ],
          "type": "tbody",
        },
      ],
      "type": "table",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
